File: cs/test/Ice/background/BackgroundControllerI.cs
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using Test;
using System.Threading;
using System.Collections.Generic;

internal class BackgroundControllerI : BackgroundControllerDisp_
{
    public override void pauseCall(string opName, Ice.Current current)
    {
        _m.Lock();
        try
        {
            _pausedCalls.Add(opName);
        }
        finally
        {
            _m.Unlock();
        }
    }

    public override void resumeCall(string opName, Ice.Current current)
    {
        _m.Lock();
        try
        {
            _pausedCalls.Remove(opName);
            _m.NotifyAll();
        }
        finally
        {
            _m.Unlock();
        }
    }

    internal void checkCallPause(Ice.Current current)
    {
        _m.Lock();
        try
        {
            while(_pausedCalls.Contains(current.operation))
            {
                _m.Wait();
                break;
            }
        }
        finally
        {
            _m.Unlock();
        }
    }

    public override void holdAdapter(Ice.Current current)
    {
        _adapter.hold();
    }

    public override void resumeAdapter(Ice.Current current)
    {
        _adapter.activate();
    }

    public override void initializeException(bool enable, Ice.Current current)
    {
        _configuration.initializeException(enable ? new Ice.SocketException() : null);
    }

    public override void readReady(bool enable, Ice.Current current)
    {
        _configuration.readReady(enable);
    }

    public override void readException(bool enable, Ice.Current current)
    {
        _configuration.readException(enable ? new Ice.SocketException() : null);
    }

    public override void writeReady(bool enable, Ice.Current current)
    {
        _configuration.writeReady(enable);
    }

    public override void writeException(bool enable, Ice.Current current)
    {
        _configuration.writeException(enable ? new Ice.SocketException() : null);
    }

    internal BackgroundControllerI(Ice.ObjectAdapter adapter)
    {
        _adapter = adapter;
        _configuration = Configuration.getInstance();
    }

    private Ice.ObjectAdapter _adapter;
#if COMPACT
    private IceInternal.HashSet<string> _pausedCalls = new IceInternal.HashSet<string>();
#else
    private HashSet<string> _pausedCalls = new HashSet<string>();
#endif
    private Configuration _configuration;
    private readonly IceUtilInternal.Monitor _m = new IceUtilInternal.Monitor();
}

File: java/test/Ice/background/BackgroundControllerI.java
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************
package test.Ice.background;

import test.Ice.background.Test._BackgroundControllerDisp;

class BackgroundControllerI extends _BackgroundControllerDisp
{
    synchronized public void
    pauseCall(String opName, Ice.Current current)
    {
        _pausedCalls.add(opName);
    }

    synchronized public void
    resumeCall(String opName, Ice.Current current)
    {
        _pausedCalls.remove(opName);
        notifyAll();
    }

    synchronized public void
    checkCallPause(Ice.Current current)
    {
        while(_pausedCalls.contains(current.operation))
        {
            try
            {
                wait();
                break;
            }
            catch(java.lang.InterruptedException ex)
            {
            }
        }
    }

    public void
    holdAdapter(Ice.Current current)
    {
        _adapter.hold();
    }
    
    public void
    resumeAdapter(Ice.Current current)
    {
        _adapter.activate();
    }
    
    public void
    initializeSocketStatus(int status, Ice.Current current)
    {
        _configuration.initializeSocketStatus(status);
    }

    public void
    initializeException(boolean enable, Ice.Current current)
    {
        _configuration.initializeException(enable ? new Ice.SocketException() : null);
    }

    public void
    readReady(boolean enable, Ice.Current current)
    {
        _configuration.readReady(enable);
    }

    public void
    readException(boolean enable, Ice.Current current)
    {
        _configuration.readException(enable ? new Ice.SocketException() : null);
    }
    
    public void
    writeReady(boolean enable, Ice.Current current)
    {
        _configuration.writeReady(enable);
    }

    public void
    writeException(boolean enable, Ice.Current current)
    {
        _configuration.writeException(enable ? new Ice.SocketException() : null);
    }
    
    public
    BackgroundControllerI(Configuration configuration, Ice.ObjectAdapter adapter)
    {
        _adapter = adapter;
        _configuration = configuration;
    }

    final private Ice.ObjectAdapter _adapter;
    final private java.util.Set<String> _pausedCalls = new java.util.HashSet<String>();
    final private Configuration _configuration;
}
